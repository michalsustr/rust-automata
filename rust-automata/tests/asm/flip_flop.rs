use rust_automata::*;
use std::process::ExitCode;

#[state_machine(
    states(Flip, Flop),
    transitions(
        (Flip) -> (Flop),
        (Flop) -> (Flip),
    ),
    generate_structs(true),
    derive(Debug, PartialEq),
)]
pub struct FlipFlop;


fn main() -> ExitCode {
    let mut m = StateMachine::new(FlipFlop, Flip);
    m.step();
    // If you remove one step call, the generated assembly will be still optimal.
    m.step();

    if m.state().is_flip() {
        ExitCode::SUCCESS
    } else {
        ExitCode::FAILURE
    }
}
