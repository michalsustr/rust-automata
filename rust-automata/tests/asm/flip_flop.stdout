flip_flop::main:
 xor     eax, eax
 ret
